### PR TITLE
enable UAA microservices w/o eureka

### LINKED ISSUE #8967: Improve docker-compose for non-eureka microservices
##### **Overview of the feature request**

I talked with @saturnism during JHipster Conf in Paris about our microservice experience when working with kubernetes. In a nutshell, using K8s native service discovery, ingress etc, we can get rid of Zuul and Eureka. This is mainly interesting for K8s deployments but doesn't help so much for developer experience.

In other words, you can use Minikube for local development. But usually, this comes with certain limitations for the developer. Running like 10 java microservices inside a VM for Minikube will lack performance. When speaking for myself, my workflow is developing using docker-compose and use kubernetes in production. But our docker-compose isn't aware of that level of service discovery and ingress routing like k8s does, but there are approaches to achieve this.

So I am thinking of 2 changes for our docker-compose generator:

1. create a small project called "jhipster-proxy", which is an extension on nginx (or maybe traefik) in order to pass service names, so they are resolved similarly to what Zuul does automatically, or what our enhanced ingress generator does. In more details, if we have a "gateway", "fooservice" and "barservice", we should route the services over a reverse proxy, and all requests on '/' to the gateway and expose the proxy instead of the gateway to the host port (which usually is 8080). 

2. find a solution how to make the services being resolved with the same hostname like inside k8s or using service discovery. docker-compose already supports hostname resolution by the name of the services, but currently we generate names like "fooservice-app" instead of "fooservice". So we could just rename these names or use some option for docker-compose to make the default app names work.

I am currently working on a prototype at work to solve these 2 things and would love to PR them as soon as they start to work.

-   [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by xetys**
this PR enhances the experience when using microservices (in particular
UAA) without eureka in docker-compose and kubernetes

closes #8967 and #9027

-   [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by xetys** [comment]
Hey, 

here is my first approach how to make UAA based microservices work again when running without eureka. In a nutshell, I have the following issue. When we use eureka, we always can contact the services by their names, say `http://userservice/api/...`. With eureka as service discovery, we can fetch the IP and port implicitly. If we don't use eureka, we should:

1. pass the ports to all calls. For example, in all UAA related stuff like token endpoints (`http://userservice:9999/oauth/token_key`), but also for all microservice calls for entities and so on
2. setup Zuul routes in a different way then we do with eureka

To solve this, I see two solutions:

## Port 80 and services with `RouteDetector`
1. expect all *productive* ports to be 80!. This means:
  a. in docker-compose, all **microservices** will run with `SERVER_PORT=80`, so, inside the docker-compose network, they are resolvable without a port, like `http://serviceName`
  b. in kubernetes, all services will have port 80 for the same reason (while the targetPort can stay)
2. use a `Filter` called route detector in zuul, which detects microservice call and dynamically sets up a static route.

## Full Static approach

We carry the information about microservice names and ports to all required parts (mostly it's the gateway). This is tricky, as e.g. for UAA we have only `uaaBaseName` at the moment. For microservice entities, we had an approach using needles when we were adding webpack proxies on entity generation. Here we could do something similar to zuuls static routes. Then we should involve a duplicate check, which might be somehow hacky.


IMHO, the port-80 approach makes more sense to me. In production, we don't need different ports, as they aren't exposed anyways. Both in docker-compose and in kubernetes, we can configure all services to port 80 and solve a lot of tricky problems like hacking with needles. I'd love to hear what e.g. @PierreBesson @saturnism think about it.


If no one is critical about my first approach, I'll move it to a working and ready2merge state. I can change it later anyways

**Comment by xetys** [comment]
Well, from the current state, this PR allows to run microservices (with UAA) without eureka in docker-compose and kubernetes (and consequently kubernetes-helm). In addition, I fixed the liveness probe bug introduced in #10548. While the PR has a good intension, we propably should open up the info uri for non-authenticated access or something similar.

At the current time, this PR closes #8967 #9027 and #10770

**Comment by xetys** [comment]
P.S.: the one failing test is failing anyways...not related to this PR

**Comment by pascalgrimaud** [comment]
closed / reopened to restart the build

**Comment by pascalgrimaud** [comment]
@xetys : you're the expert on this part, so I'm trusting you. Let's merge this

File: generators/server/templates/src/main/java/package/web/filter/RouteDetectorFilter.java.ejs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            String serviceName = matcher.group(1);
+
+            // test, if the service is missing in zuul routes
+            if (!zuulProperties.getRoutes().containsKey(serviceName)) {
+                try {
+                    URI uri = new URI(
+                        "http",
+                        null,
+                        serviceName,
+                        80,
```
Comment (jdubois) [inline generators/server/templates/src/main/java/package/web/filter/RouteDetectorFilter.java.ejs]: The port is hard-coded here? Isn't this going to cause trouble?

**Comment by pascalgrimaud** [comment]
@jdubois : I think you're right about port 80, and we got an issue with docker compose

Then sorry @xetys, unless we can have a fix for this, I'll revert this PR for a new release

**Pull request merged**